src/cpp/runner/main.cpp

```diff
@@ -1,151 +1,44 @@
-        input.alphas, input.means, input.icf, input.x, input.wishart, replicate_point);
-
-    return input;
-}
-
-typedef void (ITest<GMMInput, GMMOutput>::* test_member_function) (int);
-
-void call_member_function(unique_ptr<ITest<GMMInput, GMMOutput>>& ptr_to_object, const test_member_function ptr_to_member, const int arg1) {
-    (*(ptr_to_object).*(ptr_to_member))(arg1);
-}
-
-double measure_shortest_time(const double minimum_measurable_time, const int nruns, const double time_limit, unique_ptr<ITest<GMMInput, GMMOutput>>& test, const test_member_function func)
-{
-    auto min_sample = std::numeric_limits<double>::max();
-    double total_time = 0;
-    auto repeats = 1;
-    while (repeats < (1 << 30))
-    {
-        auto t1 = high_resolution_clock::now();
-        call_member_function(test, func, repeats);
-        auto t2 = high_resolution_clock::now();
-        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
-        if (current_run_time > minimum_measurable_time)
-        {
-            min_sample = std::min(min_sample, current_run_time / repeats);
-            total_time += current_run_time;
-            break;
-        }
-        repeats *= 2;
-    }
-
-    for (auto run = 1; (run < nruns) && (total_time < time_limit); run++)
-    {
-        auto t1 = high_resolution_clock::now();
-        call_member_function(test, func, repeats);
-        auto t2 = high_resolution_clock::now();
-        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
-        min_sample = std::min(min_sample, current_run_time / repeats);
-        total_time += current_run_time;
-    }
-
-    return min_sample;
-}
-
-std::string filepath_to_basename(const std::string& filepath)
-{
-    const auto last_slash_position = filepath.find_last_of("/\\");
-    const auto filename = last_slash_position == std::string::npos
-        ? filepath
-        : filepath.substr(last_slash_position + 1);
-
-    const auto dot = filename.find_last_of('.');
-    const auto basename = dot == std::string::npos
-        ? filename
-        : filename.substr(0, dot);
-
-    return basename;
-}
-
-void save_time_to_file(const string& filepath, const double objective_time, const double derivative_time)
-{
-    std::ofstream out(filepath);
-    out << std::scientific << objective_time << std::endl << derivative_time;
-    out.close();
-}
-
-void save_objective_to_file(const string& filepath, const double value)
-{
-    std::ofstream out(filepath);
-    out << std::scientific << value;
-    out.close();
-}
-
-void save_gradient_to_file(const string& filepath, const vector<double>& gradient)
-{
-    std::ofstream out(filepath);
-
-    for (const auto& i : gradient)
-    {
-        out << std::scientific << i << std::endl;
-    }
-
-    out.close();
+std::string str_toupper(std::string s) {
+    std::transform(s.begin(), s.end(), s.begin(),
+        [](const unsigned char c) { return std::toupper(c); } 
```

Comment:
Is it the case that std::toupper is still a macro?  Or wrong signature?

Reply:
Sorry I don't understand what do you mean Andrew. Could you explain more?